### PR TITLE
feat: added server variable object as reusable object

### PR DESCRIPTION
**Pull request by danielkocot**
---
title: "Server Variable Object as reusable object"
---

---

**Related issue(s):**
#707 
---
The PR targets to have Server Variable Objects as reusable objects within the Component Object.

### CODE REVIEW

**Review by magicmatatjahu — CHANGES_REQUESTED** [review]

File: spec/asyncapi.md
Comment (danielkocot) [reply spec/asyncapi.md:1504]: You're totally right.

**Review by magicmatatjahu — APPROVED** [review]
For me it looks good 🚀  We have first feature to `2.4.0`!

I'm not a maintainer of that repo, so I cannot accept it 😅  cc @derberg @fmvilas 
You should be also interest @smoya @jonaslagoni

**Comment by fmvilas** [comment]
LGTM so far 👍 

The only thing I'm missing now is the PRs on the https://github.com/asyncapi/spec-json-schemas and https://github.com/asyncapi/parser-js/ repos.

**Comment by danielkocot** [comment]
asyncapi/spec-json-schemas#174

**Comment by danielkocot** [comment]
Can anybody help me with the PR for the https://github.com/asyncapi/parser-js/ repo?

**Comment by magicmatatjahu** [comment]
@danielkocot Sure! You should write needed method for new field in this class https://github.com/asyncapi/parser-js/blob/master/lib/models/components.js like `serverVariables`,  `hasServerVariables`, `serverVariables(name)` and then write tests in the https://github.com/asyncapi/parser-js/blob/master/test/models/components_test.js file for new methods. If you will have problems please ping me :)

**Comment by danielkocot** [comment]
Last missing PR: async/parser-js#476

**Comment by derberg** [comment]
@danielkocot please have a look at examples. You propose `serverVariables` to be under `components`, but in examples, it is on the root.

**Comment by derberg** [comment]
ah 😅 servers were also under components 😄 sorry folks 🍺

**Comment by smoya** [comment]
@danielkocot we have now release branches for `2.4.0`. Would you mind changing the base branch to `2022-04-release` ? Thanks!

**Comment by smoya** [comment]
@danielkocot https://github.com/asyncapi/spec/pull/736 has been merged. Would you mind changing the base branch to `2022-04-release` ? Thanks!

File: spec/asyncapi.md
Comment (dalelane) [inline spec/asyncapi.md:1632]: is the extra comma here a typo?

```suggestion
      default: demo
```

File: spec/asyncapi.md
Comment (dalelane) [inline spec/asyncapi.md:1635]: similarly here... it's possible I'm misunderstanding something :-) 

```suggestion
      enum: [8883, 8884]
```

**Review by dalelane — APPROVED** [review]
LGTM 👍

**Comment by smoya** [comment]
@derberg can you approve the workflows so they can run? Thanks

2022-04-05 at 10 55 57

**Comment by smoya** [comment]
@fmvilas as [per your comment](https://github.com/asyncapi/spec/pull/717#issuecomment-1038940892), it looks OK to you. Would you mind checking again now that all PR's have been created?

**Review by magicmatatjahu — APPROVED** [review]
LGTM :)

File: spec/asyncapi.md
Comment (smoya) [inline spec/asyncapi.md:1621]: `mdv` markdown linter [is failing](https://github.com/asyncapi/spec/runs/5832670366?check_suite_focus=true) since you are not adding proper quotation to this. In YAML, curly brackets are use for maps, so the quotes are needed.

This will fix CI
```suggestion
      url: "{stage}.gigantic-server.com:{port}"
```

Comment (smoya) [inline spec/asyncapi.md:1621]: `mdv` markdown linter [is failing](https://github.com/asyncapi/spec/runs/5832670366?check_suite_focus=true) since you are not adding proper quotation to this. In YAML, curly brackets are use for maps, so the quotes are needed.

This will fix CI
```suggestion
      url: "{stage}.gigantic-server.com:{port}"
```

**Review by smoya — CHANGES_REQUESTED** [review]
@danielkocot there is only one little thing to fix here and we will be most probably ready to merge it soon 🚀

**Comment by smoya** [comment]
We still need the approval of @derberg and @fmvilas here.

**Review by derberg — APPROVED** [review]

**Review by fmvilas — APPROVED** [review]

**Comment by smoya** [comment]
@derberg @fmvilas @dalelane if you agree, please add the RFC-3 label and proceed with merging. Thanks!

**Comment by derberg** [comment]
done, main issue has it already too

**Comment by smoya** [comment]
All PR's have been merged. Proceeding with this PR.

**Pull request merged**
